org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/handlers/WorkspaceExecuteCommandHandler.java

```diff
@@ -0,0 +1,137 @@
+ * http://www.eclipse.org/legal/epl-v10.html
+ *
+ * Contributors:
+ *     Microsoft Corporation - initial API and implementation
+ *******************************************************************************/
+
+package org.eclipse.jdt.ls.core.internal.handlers;
+
+import java.util.Collection;
+import java.util.Set;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+import org.eclipse.core.runtime.CoreException;
+import org.eclipse.core.runtime.IConfigurationElement;
+import org.eclipse.core.runtime.IProgressMonitor;
+import org.eclipse.core.runtime.ISafeRunnable;
+import org.eclipse.core.runtime.IStatus;
+import org.eclipse.core.runtime.Platform;
+import org.eclipse.core.runtime.SafeRunner;
+import org.eclipse.core.runtime.Status;
+import org.eclipse.jdt.ls.core.internal.IDelegateCommandHandler;
+import org.eclipse.jdt.ls.core.internal.JavaLanguageServerPlugin;
+import org.eclipse.lsp4j.ExecuteCommandParams;
+
+public class WorkspaceExecuteCommandHandler {
+
+	/**
+	 * Extension point ID for the delegate command handler.
+	 */
+	private static final String EXTENSION_POINT_ID = "org.eclipse.jdt.ls.core.delegateCommandHandler";
+
+	private static final String COMMAND = "command";
+
+	private static final String CLASS = "class";
+
+	private static final String ID = "id";
+
+	private static class DelegateCommandHandlerDescriptor {
+
+		private final IConfigurationElement fConfigurationElement;
+
+		private Set<String> fCommandIds;
+
+		private IDelegateCommandHandler fDelegateCommandHandlerInstance;
+
+		public DelegateCommandHandlerDescriptor(IConfigurationElement element) {
+			fConfigurationElement = element;
+
+			IConfigurationElement[] children = fConfigurationElement.getChildren(COMMAND);
+			fCommandIds = Stream.of(children).map(c -> c.getAttribute(ID)).collect(Collectors.toSet());
+			fDelegateCommandHandlerInstance = null;
+		}
+
+		public Set<String> getCommands() {
+			return fCommandIds;
+		}
+
+		public IDelegateCommandHandler getDelegateCommandHandler() {
+			if (fDelegateCommandHandlerInstance == null) {
+				try {
+					Object extension = fConfigurationElement.createExecutableExtension(CLASS);
+					if (extension instanceof IDelegateCommandHandler) {
+						fDelegateCommandHandlerInstance = (IDelegateCommandHandler) extension;
+					} else {
+						String message = "Invalid extension to " + EXTENSION_POINT_ID + ". Must implements org.eclipse.jdt.ls.core.internal.IDelegateCommandHandler";
+						JavaLanguageServerPlugin.logError(message);
+						return null;
+					}
+				} catch (CoreException e) {
+					JavaLanguageServerPlugin.logException("Unable to create delegate command handler ", e);
+					return null;
+				}
+			}
+			return fDelegateCommandHandlerInstance;
+		}
+	}
+
+	private static Set<DelegateCommandHandlerDescriptor> fgContributedCommandHandlers;
+
+	private static Set<DelegateCommandHandlerDescriptor> getDelegateCommandHandlers() {
+		if (fgContributedCommandHandlers == null) {

```

Comment:
What we're trying to make lazy is activation of the implementation bundles for the commands. I would move the processing of the extension point into the bundle activator #start() method (which is thread safe). Since we're only interacting with the extension registry, this should be fast.
Creating the executable extension must remain at invocation time, but if we remember the handler instance, we must do it in a thread safe manner.